[Bug] [Master] NPE on Master

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.


### What happened

We found some NullPointerExceptions on the master server.
```
[INFO] 2023-08-03 10:58:00.014 +0800 org.apache.dolphinscheduler.scheduler.quartz.ProcessScheduleTask:[68] - [WorkflowInstance-0][TaskInstance-0] - scheduled fire time :Thu Aug 03 10:58:00 CST 2023, fire time :Thu Aug 03 10:58:00 CST 2023, scheduleId :5
[INFO] 2023-08-03 10:58:00.406 +0800 org.apache.dolphinscheduler.server.master.runner.MasterSchedulerBootstrap:[178] - [WorkflowInstance-0][TaskInstance-0] - Master schedule bootstrap loop command success, fetch command size: 1, cost: 2ms, current slot: 0, total slot size: 1
[ERROR] 2023-08-03 10:58:00.423 +0800 org.apache.dolphinscheduler.server.master.runner.MasterSchedulerBootstrap:[147] - [WorkflowInstance-0][TaskInstance-0] - Master handle command 6601 error 
org.apache.dolphinscheduler.server.master.exception.WorkflowCreateException: Create workflow execute runnable failed
	at org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnableFactory.createWorkflowExecuteRunnable(WorkflowExecuteRunnableFactory.java:95)
	at org.apache.dolphinscheduler.server.master.runner.MasterSchedulerBootstrap.lambda$run$0(MasterSchedulerBootstrap.java:136)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1384)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
	at java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:290)
	at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinTask.doInvoke(ForkJoinTask.java:401)
	at java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:734)
	at java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:159)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:173)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:485)
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:650)
	at org.apache.dolphinscheduler.server.master.runner.MasterSchedulerBootstrap.run(MasterSchedulerBootstrap.java:133)
Caused by: java.lang.NullPointerException: null
	at org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteContextFactory.createWorkflowInstance(WorkflowExecuteContextFactory.java:79)
	at org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteContextFactory.createWorkflowExecuteRunnableContext(WorkflowExecuteContextFactory.java:54)
	at org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnableFactory.createWorkflowExecuteRunnable(WorkflowExecuteRunnableFactory.java:81)
	... 15 common frames omitted
[INFO] 2023-08-03 10:58:31.177 +0800 org.apache.dolphinscheduler.server.master.runner.StateWheelExecuteThread:[315] - [WorkflowInstance-6581][TaskInstance-0] - [TaskInstanceKey-10444805098592:3]The task instance can retry, will retry this task instance
[INFO] 2023-08-03 10:58:31.177 +0800 org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteThreadPool:[101] - [WorkflowInstance-6581][TaskInstance-0] - Submit state event success, stateEvent: TaskStateEvent(processInstanceId=6581, taskInstanceId=null, taskCode=10444805098592, status=TaskExecutionStatus{code=1, desc='running'}, type=TASK_RETRY, key=null, channel=null, context=null)
[INFO] 2023-08-03 10:58:31.259 +0800 org.apache.dolphinscheduler.server.master.runner.WorkflowExecuteRunnable:[290] - [WorkflowInstance-6581][TaskInstance-null] - Begin to handle state event, TaskStateEvent(processInstanceId=6581, taskInstanceId=null, taskCode=10444805098592, status=TaskExecutionStatus{code=1, desc='running'}, type=TASK_RETRY, key=null, channel=null, context=null)

```


We have currently identified that the NPE is caused by the fact that this piece of code returns null.
https://github.com/apache/dolphinscheduler/blob/5ec9085113c989c06a23d0e6820627b147dad15e/dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java#L331-L335




@ruanwenjun PTAL. ref: https://github.com/apache/dolphinscheduler/pull/14544

### What you expected to happen

no NPE error 

### How to reproduce

A new cluster has been deployed through k8s with the dev branch, and MySQL is used as the database. The following steps were taken:

1. Created a shell task with the script content 'exit 10086;' and set the task to retry once.
2. Created a corresponding workflow with a serial wait execution policy.
3. Created a scheduled task to execute every minute.

After waiting for the task to execute several times, a NullPointerException error log appeared.

### Anything else

_No response_

### Version

dev

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterSchedulerBootstrap.java
+++ b/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterSchedulerBootstrap.java
@@ -40,6 +40,7 @@ import org.apache.commons.collections4.CollectionUtils;
 
 import java.util.Collections;
 import java.util.List;
+import java.util.Optional;
 
 import lombok.extern.slf4j.Slf4j;
 
@@ -132,8 +133,14 @@ public class MasterSchedulerBootstrap extends BaseDaemonThread implements AutoCl
                 commands.parallelStream()
                         .forEach(command -> {
                             try {
-                                WorkflowExecuteRunnable workflowExecuteRunnable =
+                                Optional<WorkflowExecuteRunnable> workflowExecuteRunnableOptional =
                                         workflowExecuteRunnableFactory.createWorkflowExecuteRunnable(command);
+                                if (!workflowExecuteRunnableOptional.isPresent()) {
+                                    log.warn(
+                                            "The command execute success, will not trigger a WorkflowExecuteRunnable, this workflowInstance might be in serial mode");
+                                    return;
+                                }
+                                WorkflowExecuteRunnable workflowExecuteRunnable = workflowExecuteRunnableOptional.get();
                                 ProcessInstance processInstance = workflowExecuteRunnable
                                         .getWorkflowExecuteContext().getWorkflowInstance();
                                 if (processInstanceExecCacheManager.contains(processInstance.getId())) {

--- a/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteContextFactory.java
+++ b/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteContextFactory.java
@@ -29,6 +29,8 @@ import org.apache.dolphinscheduler.server.master.registry.ServerNodeManager;
 import org.apache.dolphinscheduler.service.exceptions.CronParseException;
 import org.apache.dolphinscheduler.service.process.ProcessService;
 
+import java.util.Optional;
+
 import lombok.extern.slf4j.Slf4j;
 
 import org.springframework.beans.factory.annotation.Autowired;
@@ -50,21 +52,22 @@ public class WorkflowExecuteContextFactory {
     @Autowired
     private MasterConfig masterConfig;
 
-    public IWorkflowExecuteContext createWorkflowExecuteRunnableContext(Command command) throws Exception {
-        ProcessInstance workflowInstance = createWorkflowInstance(command);
+    public Optional<IWorkflowExecuteContext> createWorkflowExecuteRunnableContext(Command command) throws Exception {
+        Optional<ProcessInstance> workflowInstanceOptional = createWorkflowInstance(command);
+        if (!workflowInstanceOptional.isPresent()) {
+            return Optional.empty();
+        }
+        ProcessInstance workflowInstance = workflowInstanceOptional.get();
         ProcessDefinition workflowDefinition = processService.findProcessDefinition(
                 workflowInstance.getProcessDefinitionCode(), workflowInstance.getProcessDefinitionVersion());
         workflowInstance.setProcessDefinition(workflowDefinition);
 
         IWorkflowGraph workflowGraph = workflowGraphFactory.createWorkflowGraph(workflowInstance);
 
-        return new WorkflowExecuteContext(
-                workflowDefinition,
-                workflowInstance,
-                workflowGraph);
+        return Optional.of(new WorkflowExecuteContext(workflowDefinition, workflowInstance, workflowGraph));
     }
 
-    private ProcessInstance createWorkflowInstance(Command command) throws CronParseException {
+    private Optional<ProcessInstance> createWorkflowInstance(Command command) throws CronParseException {
         long commandTransformStartTime = System.currentTimeMillis();
         // Note: this check is not safe, the slot may change after command transform.
         // We use the database transaction in `handleCommand` so that we can guarantee the command will
@@ -76,10 +79,9 @@ public class WorkflowExecuteContextFactory {
             throw new RuntimeException("Slot check failed the current state: " + slotCheckState);
         }
         ProcessInstance processInstance = processService.handleCommand(masterConfig.getMasterAddress(), command);
-        log.info("Master handle command {} end, create process instance {}", command.getId(), processInstance.getId());
         ProcessInstanceMetrics
                 .recordProcessInstanceGenerateTime(System.currentTimeMillis() - commandTransformStartTime);
-        return processInstance;
+        return Optional.ofNullable(processInstance);
     }
 
     private SlotCheckState slotCheck(Command command) {

--- a/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnableFactory.java
+++ b/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnableFactory.java
@@ -19,7 +19,6 @@ package org.apache.dolphinscheduler.server.master.runner;
 
 import org.apache.dolphinscheduler.dao.entity.Command;
 import org.apache.dolphinscheduler.dao.repository.ProcessInstanceDao;
-import org.apache.dolphinscheduler.dao.repository.TaskDefinitionLogDao;
 import org.apache.dolphinscheduler.dao.repository.TaskInstanceDao;
 import org.apache.dolphinscheduler.server.master.config.MasterConfig;
 import org.apache.dolphinscheduler.server.master.exception.WorkflowCreateException;
@@ -29,6 +28,8 @@ import org.apache.dolphinscheduler.service.alert.ProcessAlertManager;
 import org.apache.dolphinscheduler.service.command.CommandService;
 import org.apache.dolphinscheduler.service.expand.CuringParamsService;
 import org.apache.dolphinscheduler.service.process.ProcessService;
+
+import java.util.Optional;
 
 import lombok.extern.slf4j.Slf4j;
 
@@ -67,20 +68,17 @@ public class WorkflowExecuteRunnableFactory {
     private MasterConfig masterConfig;
 
     @Autowired
-    private TaskDefinitionLogDao taskDefinitionLogDao;
-
-    @Autowired
     private DefaultTaskExecuteRunnableFactory defaultTaskExecuteRunnableFactory;
 
     @Autowired
     private WorkflowExecuteContextFactory workflowExecuteContextFactory;
 
-    public WorkflowExecuteRunnable createWorkflowExecuteRunnable(Command command) throws WorkflowCreateException {
+    public Optional<WorkflowExecuteRunnable> createWorkflowExecuteRunnable(Command command) throws WorkflowCreateException {
         try {
-            IWorkflowExecuteContext workflowExecuteRunnableContext =
+            Optional<IWorkflowExecuteContext> workflowExecuteRunnableContextOptional =
                     workflowExecuteContextFactory.createWorkflowExecuteRunnableContext(command);
-            return new WorkflowExecuteRunnable(
-                    workflowExecuteRunnableContext,
+            return workflowExecuteRunnableContextOptional.map(iWorkflowExecuteContext -> new WorkflowExecuteRunnable(
+                    iWorkflowExecuteContext,
                     commandService,
                     processService,
                     processInstanceDao,
@@ -90,7 +88,7 @@ public class WorkflowExecuteRunnableFactory {
                     stateWheelExecuteThread,
                     curingGlobalParamsService,
                     taskInstanceDao,
-                    defaultTaskExecuteRunnableFactory);
+                    defaultTaskExecuteRunnableFactory));
         } catch (Exception ex) {
             throw new WorkflowCreateException("Create workflow execute runnable failed", ex);
         }

--- a/dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessService.java
+++ b/dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessService.java
@@ -56,11 +56,14 @@ import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 
+import javax.annotation.Nullable;
+
 import org.springframework.transaction.annotation.Transactional;
 
 public interface ProcessService {
 
     @Transactional
+    @Nullable
     ProcessInstance handleCommand(String host,
                                   Command command) throws CronParseException, CodeGenerateUtils.CodeGenerateException;
 

--- a/dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java
+++ b/dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java
@@ -303,6 +303,7 @@ public class ProcessServiceImpl implements ProcessService {
     @Autowired
     private TriggerRelationService triggerRelationService;
     /**
+     * todo: split this method
      * handle Command (construct ProcessInstance from Command) , wrapped in transaction
      *
      * @param host    host
@@ -311,8 +312,8 @@ public class ProcessServiceImpl implements ProcessService {
      */
     @Override
     @Transactional
-    public ProcessInstance handleCommand(String host,
-                                         Command command) throws CronParseException, CodeGenerateException {
+    public @Nullable ProcessInstance handleCommand(String host,
+                                                   Command command) throws CronParseException, CodeGenerateException {
         ProcessInstance processInstance = constructProcessInstance(command, host);
         // cannot construct process instance, return null
         if (processInstance == null) {
@@ -332,6 +333,7 @@ public class ProcessServiceImpl implements ProcessService {
                 setSubProcessParam(processInstance);
                 triggerRelationService.saveProcessInstanceTrigger(command.getId(), processInstance.getId());
                 deleteCommandWithCheck(command.getId());
+                // todo: this is a bad design to return null here, whether trigger the task
                 return null;
             }
         } else {

